### PR TITLE
Support feature/nullish coalesce

### PR DESCRIPTION
**Pull request by Yi2255**
Add the new operator `NullCoalesce`

### CODE REVIEW

**Comment by saelo** [comment]
To clarify, what's the relationship between this PR and #464? Is this a replacement?

**Comment by Yi2255** [comment]
Yes, this PR is a revised version based on your feedback on [#464.](https://github.com/googleprojectzero/fuzzilli/pull/464)

File: Sources/Fuzzilli/Mutators/RuntimeAssistedMutator.swift [hunk 1]
```diff
@@ -186,6 +186,7 @@ public class RuntimeAssistedMutator: Mutator {
         case LogicalAnd = "LOGICAL_AND"
         case LogicalOr = "LOGICAL_OR"
         case LogicalNot = "LOGICAL_NOT"
+        case NullCoalesce = "NullCoalesce"
```
Comment (saelo) [inline Sources/Fuzzilli/Mutators/RuntimeAssistedMutator.swift:189]: Let's stay consistent here, e.g. `NULL_COALESCE`

File: Sources/Fuzzilli/Lifting/JavaScriptRuntimeAssistedMutatorLifting.swift [hunk 2]
```diff
@@ -389,6 +389,7 @@ struct JavaScriptRuntimeAssistedMutatorLifting {
         const OP_LOGICAL_AND = 'LOGICAL_AND';
         const OP_LOGICAL_OR = 'LOGICAL_OR';
         const OP_LOGICAL_NOT = 'LOGICAL_NOT';
+        const OP_LOGICAL_NullCoalesce = 'LOGICAL_NullCoalesce';
```
Comment (saelo) [inline Sources/Fuzzilli/Lifting/JavaScriptRuntimeAssistedMutatorLifting.swift:392]: Same here, this should probably be `OP_NULL_COALESCE`

File: Sources/Fuzzilli/Lifting/JavaScriptRuntimeAssistedMutatorLifting.swift [hunk 3]
```diff
@@ -511,6 +512,7 @@ struct JavaScriptRuntimeAssistedMutatorLifting {
           [OP_LOGICAL_AND]: (inputs) => inputs[0] && inputs[1],
           [OP_LOGICAL_OR]: (inputs) => inputs[0] || inputs[1],
           [OP_LOGICAL_NOT]: (inputs) => !inputs[0],
+          [OP_LOGICAL_NullCoalesce]: (inputs) => inputs[0] ?? inputs[1],
```
Comment (saelo) [inline Sources/Fuzzilli/Lifting/JavaScriptRuntimeAssistedMutatorLifting.swift:515]: This is good, but to be useful, you also need to create these operations. I think mainly this is useful for the ExplorationMutator which creates new instructions depending on the runtime type of variables. Concretely, what I would suggest is replacing this here: https://github.com/googleprojectzero/fuzzilli/blob/5798638831dbc3347a0aa557f14ef63ebed704b3/Sources/Fuzzilli/Lifting/JavaScriptExploreLifting.swift#L250 and here: https://github.com/googleprojectzero/fuzzilli/blob/5798638831dbc3347a0aa557f14ef63ebed704b3/Sources/Fuzzilli/Lifting/JavaScriptExploreLifting.swift#L425 to call a new `exploreNullish` function which then generates this operation. Something like:
```
function exploreNullish(v) {
    // Best thing we can do with nullish values is a NullCoalescing (??) operation.
    return new Action(OP_NULL_COALESCE, [exploredValueInput, Inputs.randomArgument()])
}
```

Then this mutator should be able to generate these new operations :)

Comment (Yi2255) [reply Sources/Fuzzilli/Lifting/JavaScriptRuntimeAssistedMutatorLifting.swift]: Thank you for your valuable advice. I did miss many details. I have modified and then pushed it.

**Comment by saelo** [comment]
Cool, thanks a lot!

**Pull request merged**